src/Passes/Source/Rules/Factory.cpp

```diff
@@ -502,6 +507,82 @@ namespace quantum
              replace_branch_positive});
     }
 
+    void RuleFactory::optimiseSelectQuantumOne()
+    {
+        auto replace_select_positive = [](Builder& builder, Value* val, Captures& cap, Replacements& replacements) {
+            auto result = cap["result"];
+            auto cond   = llvm::dyn_cast<llvm::Instruction>(cap["cond"]);
+            // Replacing result
+            auto orig_instr = llvm::dyn_cast<llvm::Instruction>(val);
+            if (orig_instr == nullptr)
+            {
+                return false;
+            }
+
+            auto                      module   = orig_instr->getModule();
+            auto                      function = module->getFunction("__quantum__qir__read_result");
+            std::vector<llvm::Value*> arguments;
+            arguments.push_back(result);
+
+            if (!function)
+            {
+                std::vector<llvm::Type*> types;
+                types.resize(arguments.size());
+                for (uint64_t i = 0; i < types.size(); ++i)
+                {
+                    types[i] = arguments[i]->getType();
+                }
+
+                auto return_type = llvm::Type::getInt1Ty(val->getContext());
+
+                llvm::FunctionType* fnc_type = llvm::FunctionType::get(return_type, types, false);
+                function                     = llvm::Function::Create(
+                    fnc_type, llvm::Function::ExternalLinkage, "__quantum__qir__read_result", module);
+            }
+
+            builder.SetInsertPoint(llvm::dyn_cast<llvm::Instruction>(val));
+            auto new_call = builder.CreateCall(function, arguments);
+            if (cond == nullptr)
+            {
+                return false;
+            }
+
+            new_call->takeName(cond);
+
+            for (auto& use : cond->uses())
+            {
+                llvm::User* user = use.getUser();
+                user->setOperand(use.getOperandNo(), new_call);
+            }
+            cond->replaceAllUsesWith(new_call);
+
+            // Deleting the previous condition and function to fetch one
+            replacements.push_back({cond, nullptr});
+            replacements.push_back({cap["one"], nullptr});
+
+            return false;
+        };
+
+        /*
+          Here is an example IR for which we want to make a match:
+
+          %1 = call %Result* @__quantum__rt__result_get_one()
+          %2 = call i1 @__quantum__rt__result_equal(%Result* %0, %Result* %1)
+          ...
+          %5 = select i1 %2, <type> %3, <type> %4
+        */
+
+        // Variations of get_one
+        auto get_one = call("__quantum__rt__result_get_one");
+        addRule(
+            {select("cond"_cap = call("__quantum__rt__result_equal", "result"_cap = _, "one"_cap = get_one), _, _),
```

Comment:
Just curiosity (feel free to ignore):
What is `"cond"_cap`, `"result"_cap`?  (I have never met such C++ constructs before)
And what is `_` in the fragment `"result"_cap = _, "one"_cap = get_one), _, _`?

Reply:
These are user literals to allow for a custom pattern matching language using native C++. All `"xxx"_cap` states is: If there is a match of the right hand side, store the match in the variable name `xxx`. The variable `_` is a shorthand notation for pattern.